pkg/ebpf/events_pipeline.go

```diff
@@ -38,6 +43,76 @@ func (t *Tracee) handleEvents(ctx gocontext.Context) {
 	t.WaitForPipeline(errcList...)
 }
 
+// Under some circumstances, tracee-rules might be slower to consume events
+// than tracee-ebpf is capable of generating them. This requires
+// tracee-ebpf to deal with this possible lag, but, at the same,
+// perf-buffer consumption can't be left behind (or important events coming
+// from the kernel might be loss, causing detection misses).
+//
+// There are 3 variables connected to this issue:
+//
+// 1) perf buffer could be increased to hold very big amount of memory
+//    pages: The problem with this approach is that the requested space,
+//    to perf-buffer, through libbpf, has to be contiguous and it is almost
+//    impossible to get very big contiguous allocations through mmap after
+//    a node is running for some time.
+//
+// 2) raising the events channel buffer to hold a very big amount of
+//    events: The problem with this approach is that the overhead of
+//    dealing with that amount of buffers, in a golang channel, causes
+//    event losses as well. It means this is not enough to relief the
+//    pressure from kernel events into perf-buffer.
+//
+// 3) create an internal, to tracee-ebpf, buffer based on the node size.
+
+// queueEvents implements an internal FIFO queue for caching events
+func (t *Tracee) queueEvents(ctx gocontext.Context, in <-chan *trace.Event) (chan *trace.Event, chan error) {
+	out := make(chan *trace.Event, 10000)
+	errc := make(chan error, 1)
+	done := make(chan bool, 1)
+
+	err := t.config.Cache.Setup()
+	if err != nil {
+		errc <- err
+		return out, errc
+	}
+
+	// receive and cache events (release pressure in the pipeline)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				done <- true
+				return
+			case event := <-in:
+				if event != nil {
+					t.config.Cache.Enqueue(event) // may block if queue is full
```

Comment:
we may block because of cond.Wait() inside the memory queue implementation BUT that is not a big deal since events will always happen and they will always unblock this (if done has to be processed).